R/coh_matching.R

```diff
@@ -0,0 +1,204 @@
+#' data("cohortdata")
+#'
+#' # assign vaccination status
+#' cohortdata$vaccine_status <- set_status(
+#'   data = cohortdata,
+#'   col_names = c("vaccine_date_1", "vaccine_date_2"),
+#'   status = c("v", "u")
+#' )
+#'
+#' # match cohort
+#' matched_cohort <- match_cohort(data = cohortdata,
+#'   status_vacc_col = "vaccine_status",
+#'   nearest = "age",
+#'   exact = "sex",
+#'   caliper = c(age = 1)
+#' )
+#'
+#' # view matched data
+#' head(matched_cohort)
+#' @export
+match_cohort <- function(data,
+                         status_vacc_col,
+                         exact = NULL,
+                         nearest = NULL,
+                         caliper = NULL) {
+
+  # input checking
+  checkmate::assert_data_frame(
+    data,
+    min.rows = 1, min.cols = 1
+  )
+  checkmate::assert_character(status_vacc_col,
+    any.missing = FALSE, min.len = 1
+  )
+  checkmate::assert_names(
+    names(data),
+    must.include = c(status_vacc_col)
+  )
+
+  # `exact` and `nearest` cannot be NULL. At least one must be provided
+  stopifnot(
+    "`exact` and `nearest` cannot be NULL. At least one must be provided" =
+      (!missing(nearest) || !missing(exact))
+  )
+
+  # checks for `nearest`
+  if (!is.null(nearest)) {
+    stopifnot(
+      "`caliper` must be provided together with `nearest`" =
+        (!missing(caliper))
+    )
+    checkmate::assert_character(nearest,
+      any.missing = FALSE, min.len = 1
+    )
+    checkmate::assert_names(
+      names(data),
+      must.include = nearest
+    )
+    checkmate::assert_numeric(
+      caliper,
+      len = length(nearest), names = "named"
+    )
+    checkmate::assert_names(
+      nearest,
+      must.include = names(caliper)
+    )
+  }
+  # checks for `exact`. Not else, both can be non-NULL
+  if (!is.null(exact)) {
+    checkmate::assert_character(exact,
+      any.missing = FALSE, min.len = 1
+    )
+    checkmate::assert_names(
+      names(data),
+      must.include = exact
+    )
+  }
+
+  # if `nearest` is NULL and `caliper` is provided, caliper is ignored
+  if (is.null(nearest) && !is.null(caliper)) {
+    warning("`caliper` ignored caused by `nearest` not provided")
+    #caliper forced to null to avoid unexpected behavior
+    caliper <- NULL
+  }
+
+  #Formula
+  variables <- c(exact, nearest)
+  formula <- paste0(status_vacc_col, " ~ ")
+  for (v in seq_along(variables)) {
+    if (v == 1) {
+      formula <- paste0(formula, variables[v])
+    } else {
+      formula <- paste0(formula, " + ", variables[v])
+    }
+  }
+  formula_eval <- eval(parse(text = formula))
+  data[[status_vacc_col]] <- as.factor(data[[status_vacc_col]])
+
+  #Matching
+  matchit <- MatchIt::matchit(
+    formula_eval,
+    data = data,
+    method = "nearest",
+    ratio = 1,
+    exact = exact,
+    nearest = nearest,
+    caliper = caliper,
+    distance = "glm"
+  )
+  match <- MatchIt::match.data(matchit, distance = "prop.score")
+  names(match) <- gsub(x = names(match),
```

Comment:
Candidate code to develop a function in dev-utils

Reply:
I just opened a [new branch](https://github.com/epiverse-trace/vaccineff/tree/dev-utils) to work on dev-utils